### PR TITLE
annotations support

### PR DESCRIPTION
**Pull request by crazy-max**
fixes https://github.com/docker/metadata-action/issues/332
fixes https://github.com/docker/metadata-action/issues/355
~needs #353~
closes #351 

alternative to https://github.com/docker/metadata-action/pull/351 that sets an output and a bake definition for annotations: https://github.com/moby/buildkit/blob/master/docs/annotations.md

### CODE REVIEW

File: action.yml [hunk 1]
```diff
@@ -44,12 +44,28 @@ outputs:
...
+    description: 'Generated manifest level annotations'
+  annotations-manifest-descriptor:
+    description: 'Generated manifest descriptor level annotations'
   json:
     description: 'JSON output of tags and labels'
   bake-file-tags:
     description: 'Bake definition file with tags'
   bake-file-labels:
     description: 'Bake definition file with labels'
+  bake-file-annotations-index:
```
Comment (tonistiigi) [inline action.yml:61]: This 4 separate files is kind of weird in here. How would the user know which is the correct one to use? It looks like we don't know which one is correct and are pushing for the user to make the mistake for us.

Maybe this makes more sense with:
- there is only one annotation file
- we pick what is the correct default
- there is some optional override to choose non-default (or multiple?) if user really wants to mess it up.

Comment (crazy-max) [reply action.yml:61]: Yes we could provide a `bake-file-annotations` output that would default to annotations at manifest level like BuildKit does as default https://github.com/moby/buildkit/blob/master/docs/annotations.md#image-annotations

> there is some optional override to choose non-default (or multiple?) if user really wants to mess it up.

We could alter the `bake-file-annotations` using an env var like `DOCKER_METADATA_ANNOTATIONS_FORMAT` that would default to `manifest` and user could pass multiple values like `DOCKER_METADATA_ANNOTATIONS_FORMAT=manifest,index`:

```yaml
      -
        name: Docker meta
        uses: docker/metadata-action@v5
        with:
          images: name/app
      -
        name: Build
        uses: docker/bake-action@v3
        with:
          files: |
            ./docker-bake.hcl
            ${{ steps.meta.outputs.bake-file-tags }}
            ${{ steps.meta.outputs.bake-file-annotations }}
          targets: build
        env:
          DOCKER_METADATA_ANNOTATIONS_FORMAT: manifest,index
```

WDYT?

Comment (crazy-max) [reply action.yml]: Updated: https://github.com/moby/buildkit/blob/master/docs/annotations.md

**Review by tonistiigi — APPROVED** [review]

**Comment by aerosol** [comment]
Hey guys, this is causing invalid json carried over in steps metadata, an example that just broke our build pipeline:

```json
(...) ,"annotations":["manifest:org.opencontainers.image.created=2023-11-30T12:46:22.687Z",manifest:org.opencontainers.image.description=Simple, open source, lightweight (< 1 KB) and privacy-friendly web analytics alternative to Google Analytics.,manifest:org.opencontainers.image.licenses=AGPL-3.0,manifest:org.opencontainers.image.revision=7bf1e2a6ed54b88e475898a18f9690bc5f690de4,manifest:org.opencontainers.image.source=https://github.com/plausible/analytics,manifest:org.opencontainers.image.title=analytics,manifest:org.opencontainers.image.url=https://github.com/plausible/analytics,"manifest:org.opencontainers.image.version=r2023.11.30-265"]}
```

Clearly there's some JSON encoding missing there.

**Pull request merged**